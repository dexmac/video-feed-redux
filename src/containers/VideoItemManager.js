import React, { Component } from "react";
import { connect } from "react-redux";
import PropTypes from "prop-types";
import * as actions from "../store/item/actions";
import store from "../store/store";
import VideoItem from "../components/VideoItem";
import { VideoDisplayState } from "../constants/constants";
import { ERR_LOADING_VIDEO_PREFIX } from "../constants/constants";

/**
 * Handles loading and displaying of video
 */
export class VideoItemManager extends Component {
  /**
   * Load the video item when component mounts
   */
  componentWillMount() {
    store.dispatch(actions.loadVideoAction(this.props.videoItem));
  }

  /**
   * Dispatch a loading success action when ReactPlayer successfully loads video
   * @param item
   */
  onVideoLoaded = item => {
    store.dispatch(actions.videoLoadingSuccess(item));
  };

  /**
   * Dispatch a loading error action - when ReactPlayer returns an error loading the video
   * @param error
   * @param item
   */
  onVideoError = (error, item) => {
    store.dispatch(actions.videoLoadingError(error, item));
  };

  render() {
    const VideoPlayer = (
      <VideoItem
        item={this.props.videoItem}
        onVideoError={this.onVideoError}
        onVideoLoaded={this.onVideoLoaded}
      />
    );

    let VideoItemDisplay;

    switch (this.props.videoDisplayState) {
      case VideoDisplayState.INIT:
      case VideoDisplayState.FETCHING:
      case VideoDisplayState.FETCHED:
        VideoItemDisplay = VideoPlayer;
        break;
      case VideoDisplayState.ERROR:
        VideoItemDisplay = (
          <div className="video-load-error-text">
            {ERR_LOADING_VIDEO_PREFIX} : {this.props.errorString}
          </div>
        );
        break;
      default:
        VideoItemDisplay = <div>ERROR: VideoItemManager - Unknown State</div>;
    }

    return <div>{VideoItemDisplay}</div>;
  }
}

VideoItemManager.propTypes = {
  videoItem: PropTypes.object
};

function mapStateToProps(state) {
  return {
    item: state.item.item,
    videoDisplayState: state.item.videoDisplayState,
    errorString: state.item.errorString
  };
}

export default connect(mapStateToProps)(VideoItemManager);
