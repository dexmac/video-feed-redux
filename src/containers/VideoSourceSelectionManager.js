import React, { Component } from "react";
import PropTypes from "prop-types";
import VideoSourceSelect from "../components/VideoSourceSelect";
import { connect } from "react-redux";
import store from "../store/store";
import { handleSourceSelectionChange } from "../store/sourceSelect/actions";

/**
 * Handles video source selection and filtering of videos based on types
 */
export class VideoSourceSelectionManager extends Component {
  actionOnSelection = selectedSources => {
    store.dispatch(handleSourceSelectionChange(selectedSources));
  };

  render() {
    const selectedValues = this.props.selectedSources
      ? this.props.selectedSources.map(source => source.value)
      : [];

    return (
      <VideoSourceSelect
        selectedValues={selectedValues}
        actionOnSelection={this.actionOnSelection}
      />
    );
  }
}

VideoSourceSelect.propTypes = {
  selectedSources: PropTypes.array
};

export function mapStateToProps(state) {
  return {
    selectedSources: state.sourceSelect.selectedSources
  };
}

export default connect(mapStateToProps)(VideoSourceSelectionManager);
