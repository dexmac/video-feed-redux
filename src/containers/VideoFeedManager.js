import React, { Component } from "react";
import { connect } from "react-redux";
import { fetchFeedItemsAction } from "../store/feed/actions";
import VideoSourceSelectionManager from "../containers/VideoSourceSelectionManager";
import VideoItemManager from "../containers/VideoItemManager";
import { VideoDisplayState } from "../constants/constants";
import store from "../store/store";

/**
 * Fetches and displays multiple video items as a single video feed
 */
export class VideoFeedManager extends Component {
  /**
   * Fetch videos with the selected sources when video source selection changes / on startup
   * @param sources
   */
  fetchVideos = sources => {
    store.dispatch(fetchFeedItemsAction(sources));
  };

  /**
   * Video feed component entry point -
   * Fetching feeds from all sources (URL, facebook & youtube)
   */
  componentDidMount() {
    this.fetchVideos();
  }

  render() {
    let Feed;

    switch (this.props.feedDisplayState) {
      case VideoDisplayState.INIT:
      case VideoDisplayState.FETCHING:
        Feed = <div className="loader" />;
        break;
      case VideoDisplayState.FETCHED:
        Feed = (
          <div className="video-feed-container">
            <VideoSourceSelectionManager fetchVideos={this.fetchVideos} />
            <div className="video-feed">
              {this.props.items.map((item, i) => {
                return <VideoItemManager videoItem={item} key={i} />;
              })}
            </div>
          </div>
        );
        break;
      case VideoDisplayState.ERROR:
        Feed = (
          <div className="video-load-error-text">{this.props.errorString}</div>
        );
        break;
      default:
        Feed = <div />;
    }

    return Feed;
  }
}

function mapStateToProps(state) {
  return {
    items: state.feed.items,
    feedDisplayState: state.feed.feedDisplayState,
    errorString: state.feed.errorString
  };
}

export default connect(mapStateToProps)(VideoFeedManager);
