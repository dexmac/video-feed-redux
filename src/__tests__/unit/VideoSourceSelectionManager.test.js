import React from "react";
import jest from "jest-mock";
import configureMockStore from "redux-mock-store";
const mockStore = configureMockStore();
import { VideoSourceSelectionManager } from "../../containers/VideoSourceSelectionManager";
import { mapStateToProps } from "../../containers/VideoSourceSelectionManager";

describe("VideoSourceSelectionManager", () => {
  let props;
  let mountedVideoSourceSelectionManager;

  /**
   * Clearing props before each test
   */
  beforeEach(() => {
    props = {
      selectedSources: ["url"]
    };
    mountedVideoSourceSelectionManager = undefined;
  });

  /**
   * Shallow rendering vs. snapshot test
   */
  it("renders correctly && matches snapshot", () => {
    const tree = shallow(<VideoSourceSelectionManager />);
    expect(tree).toMatchSnapshot();
  });

  /**
   * Mounts the component using Enzyme if it is not mounted yet
   * @returns {object} - returns the mounted component
   */
  const videoSourceSelect = () => {
    if (!mountedVideoSourceSelectionManager) {
      mountedVideoSourceSelectionManager = mount(
        <VideoSourceSelectionManager {...props} />
      );
    }
    return mountedVideoSourceSelectionManager;
  };

  describe("when `fetchVideos` method is defined and a source has been selected", () => {
    beforeEach(() => {
      props.selectedSources = ["url"];
    });

    it("calls the actionOnSelection method & checks mapStateToProps", () => {
      const videoSelect = videoSourceSelect();
      // Need to find better way to mock and call method from actual source
      videoSelect.instance().actionOnSelection([]);
      const initialState = { sourceSelect: props };
      expect(mapStateToProps(initialState)).toEqual(props);
    });
  });
});
