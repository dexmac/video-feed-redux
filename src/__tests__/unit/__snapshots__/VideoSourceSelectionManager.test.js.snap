// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VideoSourceSelectionManager renders correctly && matches snapshot 1`] = `
<VideoSourceSelect
  actionOnSelection={[Function]}
  selectedValues={Array []}
/>
`;
